content-src/selectors/selectors.js

```diff
@@ -1,144 +1,68 @@
-);
-
-const selectTopSites = module.exports.selectTopSites = createSelector(
-  [
-    state => state.TopSites
-  ],
-  TopSites => Object.assign({}, TopSites, {
-    rows: dedupe.one(TopSites.rows
-      // Add first run stuff to the end if init has already happened
-      .concat(TopSites.init ? firstRunData.TopSites : []))
-  })
-);
-
 module.exports.selectNewTabSites = createSelector(
   [
-    selectWeightedHighlights,
     state => state.Experiments.values.weightedHighlights,
-    selectTopSites,
-    state => state.History,
-    selectSpotlight,
+    state => (state.Experiments.values.weightedHighlights ? state.WeightedHighlights : selectSpotlight(state)),
+    state => state.TopSites,
+    state => (state.Experiments.values.weightedHighlights ? Object.assign({}, state.History, {rows: []}) : state.History),
     state => state.Experiments
   ],
-  (WeightedHighlights, prefWeightedHighlights, TopSites, History, Spotlight, Experiments) => {
-    // Remove duplicates
-    // Note that we have to limit the length of topsites, spotlight so we
-    // don't dedupe against stuff that isn't shown
-    let [topSitesRows, spotlightRows] = dedupe.group([TopSites.rows.slice(0, TOP_SITES_LENGTH), Spotlight.rows]);
-
-    // Find the index of the recommendation. If we have an index, find that recommendation and put it
-    // in the third highlights spot
-    let recommendation = spotlightRows.findIndex(element => element.recommended);
-    if (recommendation >= 0) {
-      spotlightRows.splice(2, 0, spotlightRows.splice(recommendation, 1)[0]);
-    }
-    spotlightRows = spotlightRows.slice(0,
-      prefWeightedHighlights ? WEIGHTED_HIGHLIGHTS_LENGTH :
-      SPOTLIGHT_DEFAULT_LENGTH);
-    const historyRows = dedupe.group([
-      topSitesRows,
-      spotlightRows,
-      History.rows])[2];
-
-    if (prefWeightedHighlights) {
-      spotlightRows = selectAndDedupe({
-        dedupe: topSitesRows,
-        sites: WeightedHighlights.rows,
-        max: WEIGHTED_HIGHLIGHTS_LENGTH,
+  (isNewHighlights, Highlights, TopSites, History, Experiments) => {
+    const [topSitesRows, highlightsRows, historyRows] = selectAndDedupe([
+      {
+        sites: TopSites.rows,
+        max: TOP_SITES_LENGTH,
+        defaults: firstRunData.TopSites
+      },
+      {
+        sites: isNewHighlights ? assignImageAndBackgroundColor(Highlights.rows) : Highlights.rows,
+        max: isNewHighlights ? WEIGHTED_HIGHLIGHTS_LENGTH : SPOTLIGHT_DEFAULT_LENGTH,
         defaults: assignImageAndBackgroundColor(firstRunData.Highlights)
-      });
+      },
+      {sites: History.rows}
+    ]);
+
+    if (!isNewHighlights) {
+      // Find the index of the recommendation. If we have an index, find that recommendation and put it
+      // in the third highlights spot
+      const recommendation = highlightsRows.findIndex(element => element.recommended);
+      if (recommendation >= 0) {
+        highlightsRows.splice(2, 0, highlightsRows.splice(recommendation, 1)[0]);
+      }
```

Comment:
you know what, just remove this while you're here - we're not showing recommendations and the logic will most likely change for showing recommendations anyways so feel free to take this out